packages/webpack/lib/utils/resolvable.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+exports.Resolvable = class Resolvable {
```

Comment:
I like this approach, but I think this file could be much simpler as the promise class already has a queueing behaviour.

```
exports.Resolvable = class Resolvable {
  constructor() {
    this.reset = this.reset.bind(this);
    this.reset();
  }

  reset() {
    const promise = new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });

    this.then = promise.then.bind(promise);
    this.catch = promise.catch.bind(promise);
  }
};
```

I've ignored the `executor` logic as it is not used...

Reply:
Nice one!

Unfortunately, it leaves us with 'orphaned' listeners - imagine a scenario as follows:
* compilation starts (`rslvbl.reset()`)
* listener subscribes  (`rslvbl.then()`)
* compilation restarts (`rslvbl.reset()`)
* compilation finishes (`rslvbl.resolve('foo')`) 
* listener is still unresolved... 😒

How about this implementation (including `executor`):

```javascript
class Resolvable {
  constructor(executor) {
    this.reset = () => {
      const { resolve: oldResolve, reject: oldReject } = this;
      const promise = new Promise((newResolve, newReject) => {
        this.resolve = (value) => {
          if (oldResolve) oldResolve(value);
          newResolve(value);
        };
        this.reject = (reason) => {
          if (oldReject) oldReject(reason);
          newReject(reason);
        };
      });
      this.then = promise.then.bind(promise);
      this.catch = promise.catch.bind(promise);
    };
    this.reset();
    if (executor) {
      new Promise(executor).then(this.resolve, this.reject);
    }
  }
}
```